Improve performance when listing columns in Iceberg

shohamyamin opened the pull request:
# Improve Performance for Retrieving Tables Metadata from Iceberg Catalog

## Overview

This PR addresses a performance issue encountered when querying column metadata from the Iceberg catalog in Trino (Issue #[23468](https://github.com/trinodb/trino/issues/23468)). The primary concern is that requests to retrieve table metadata are executed sequentially for each table, which significantly impacts query performance when dealing with a large number of tables.

## Description

The issue manifests when executing the following query on the Iceberg catalog:

```sql
SELECT * FROM iceberg.information_schema.columns;
```

In environments with a large number of tables, the query’s response time increases considerably due to the sequential execution of catalog requests.

### Changes Made

To improve the performance, this PR introduces parallelism in the metadata retrieval process. The old code executed catalog requests sequentially for each table, which resulted in longer execution times. The updated code utilizes `CompletableFuture` to handle requests asynchronously, thereby reducing the overall execution time.

The updated implementation leverages `CompletableFuture` to process tables concurrently, significantly reducing the latency in retrieving column metadata.

## Additional Context and Related Issues

Fixes #23468

### Test Setup & Observations:
 I checked the code changes and get faster query in over x4 faster then without the proposed change
- Trino Version: 463 and trino with the code in master branch with the proposed change lets called it trino 464-SNAPSHOT
- Nessie Version: 0.94.4 (In memory)
- REST Catalog based on JDBC (e.g., Tabular REST Catalog)
- Testing environment: Docker Compose with MinIO for Iceberg storage and Jaeger for tracing.
- Hardware: Intel Core i9 processor, 32GB RAM

### Performance Comparison

As you can see in this table we get around 70% reduce in time for retrieving the columns

| **Number of Tables** | **Trino Version** | **Catalog Setup** | **Query Execution Time (s)** | **Performance Increase (%)** |
|-----------------------|-------------------|-------------------|-------------------------------|------------------------------|
| **10**                | Trino 463         | REST              | 0.691                         |                              |
|                       |                   | Nessie            | 1.07                          |                              |
|                       | Trino 464-SNAPSHOT         | REST              | 0.677                         |  2.02%    |
|                       |                   | Nessie            | 0.982                         | 8.23%       |
|-----------------------|-------------------|-------------------|-------------------------------|------------------------------|
| **100**               | Trino 463         | REST              | 1.24                          |                              |
|                       |                   | Nessie            | 1.89                          |                              |
|                       | Trino 464-SNAPSHOT         | REST              | 0.772                         | 37.90%      |
|                       |                   | Nessie            | 1.29                          | 31.59%       |
|-----------------------|-------------------|-------------------|-------------------------------|------------------------------|
| **1,000**             | Trino 463         | REST              | 3.06                          |                              |
|                       |                   | Nessie            | 6.55                          |                              |
|                       | Trino 464-SNAPSHOT         | REST              | 1.17                          | 61.59%      |
|                       |                   | Nessie            | 2.42                          | 63.01%      |
|-----------------------|-------------------|-------------------|-------------------------------|------------------------------|
| **10,000**            | Trino 463         | REST              | 11.38                         |                              |
|                       |                   | Nessie            | 29.59                         |                              |
|                       | Trino 464-SNAPSHOT         | REST              | 3.41                          | 70.04%    |
|                       |                   | Nessie            | 7.70                          | 73.00%     |
|-----------------------|-------------------|-------------------|-------------------------------|------------------------------|
| **100,000**             | Trino 463         | REST              | 105                           |                              |
|                       |                   | Nessie            | 255                           |                              |
|                       | Trino 464-SNAPSHOT         | REST              | 16.96                         | 83.86%        |
|                       |                   | Nessie            | 34.72                         | 86.38%        |

- Each table in this  comparison has 10 columns
- I ran each test several times, and the results were largely consistent.

## Release Notes

```md
## Iceberg
* Improve performance when listing columns. 
```

alonahmias commented:
Thank you for this PR! This solution makes a huge difference in reducing the time needed to retrieve all available columns in Iceberg. The performance improvement is very noticeable, especially with larger datasets

raunaqmorarka commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -865,23 +867,30 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
-                        }
-                    }
+                    List<CompletableFuture<Void>> futures = remainingTables.stream()
+                            .map(tableName -> CompletableFuture.runAsync(() -> {
```
We generally don't use the common fork join pool in Trino. We should inject a dedicated executor which is bounded at some number and use that.
GET_METADATA_BATCH_SIZE is 1000, I don't think we want that many parallel fetches, something like 8 parallel fetches is probably enough.

raunaqmorarka replied:
See the way BoundedExecutor is setup in https://github.com/trinodb/trino/pull/24110/files#diff-e72d63b77858985902ad4bb32d26e11cfaab0470c78821ce5416c2c97da4deecR222

shohamyamin replied:
@raunaqmorarka Yes I was exploring how this was implemented there earlier. I’ve just updated the code. Could you please review it again? This time, I’ve incorporated the `metadataParallelism` configuration parameter and switched to using a `FixedThreadPool` for managing threads.

ebyhr commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataFactory.java:
```diff
@@ -79,6 +81,6 @@ public IcebergMetadata create(ConnectorIdentity identity)
                 metastoreFactory,
                 addFilesProcedureEnabled,
-                allowedExtraProperties);
+                allowedExtraProperties, metadataParallelism);
```
Put `metadataParallelism` on a new line.

ebyhr commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/BaseTrinoCatalogTest.java:
```diff
@@ -120,7 +120,7 @@ public void testNonLowercaseNamespace()
                     Optional.empty(),
                     false,
-                    _ -> false);
+                    _ -> false, 8);
```
Put `8` on a new line.

ebyhr commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestTrinoGlueCatalog.java:
```diff
@@ -139,7 +139,7 @@ public void testNonLowercaseGlueDatabase()
                     Optional.empty(),
                     false,
-                    _ -> false);
+                    _ -> false, 8);
```
Put `8` on a new line.

ebyhr commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/nessie/TestTrinoNessieCatalog.java:
```diff
@@ -189,7 +189,7 @@ public void testNonLowercaseNamespace()
                     Optional.empty(),
                     false,
-                    _ -> false);
+                    _ -> false, 8);
```
Put `8` on a new line.

ebyhr commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/rest/TestTrinoRestCatalog.java:
```diff
@@ -127,7 +127,7 @@ public void testNonLowercaseNamespace()
                     Optional.empty(),
                     false,
-                    _ -> false);
+                    _ -> false, 8);
```
Put `8` on a new line.

ebyhr commented on plugin/trino-iceberg/src/test/java/org/apache/iceberg/snowflake/TestTrinoSnowflakeCatalog.java:
```diff
@@ -105,15 +105,15 @@ private static void testTableSetup()
+                                )
+                                 EXTERNAL_VOLUME = '%s'
+                                 CATALOG = 'SNOWFLAKE'
+                                 BASE_LOCATION = '%s/'""".formatted(TpchTable.NATION.getTableName(), SNOWFLAKE_S3_EXTERNAL_VOLUME, TpchTable.NATION.getTableName()));
```
Revert unrelated change.

ebyhr commented on plugin/trino-iceberg/src/test/java/org/apache/iceberg/snowflake/TestTrinoSnowflakeCatalog.java:
```diff
@@ -122,14 +122,14 @@ REGIONKEY NUMBER(38,0),
+                                )
+                                 EXTERNAL_VOLUME = '%s'
+                                 CATALOG = 'SNOWFLAKE'
+                                 BASE_LOCATION = '%s/'""".formatted(TpchTable.REGION.getTableName(), SNOWFLAKE_S3_EXTERNAL_VOLUME, TpchTable.REGION.getTableName()));
```
Put `8` on a new line.

ebyhr commented on plugin/trino-iceberg/src/test/java/org/apache/iceberg/snowflake/TestTrinoSnowflakeCatalog.java:
```diff
@@ -223,7 +223,7 @@ public void testNonLowercaseNamespace()
                 Optional.empty(),
                 false,
-                _ -> false);
+                _ -> false, 8);
```
Put `8` on a new line.

ebyhr commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -1177,8 +1203,7 @@ public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, Con
             if (!replace && fileSystem.listFiles(location).hasNext()) {
-                throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, format("" +
-                        "Cannot create a table on a non-empty location: %s, set 'iceberg.unique-table-location=true' in your Iceberg catalog properties " +
+                throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, format("Cannot create a table on a non-empty location: %s, set 'iceberg.unique-table-location=true' in your Iceberg catalog properties " +
```
Revert unrelated change.

ebyhr commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -430,6 +439,16 @@ public IcebergMetadata(
+            this.executorService = Executors.newFixedThreadPool(metadataParallelism, new ThreadFactoryBuilder()
+                    .setNameFormat("iceberg-metadata-pool-%d")
+                    .setDaemon(true) // Optional: Makes threads daemon threads.
+                    .build());
```
Why do we initialize `Executor` in `IcebergMetadata` instead of providing it in `IcebergModule`?

shohamyamin replied:
I have just moved that to `IcebergMetadataFactory` and use the already existing `createIcebergMetadataExecutor` that is located in `IcebergModule`

ebyhr commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/BaseTrinoCatalogTest.java:
```diff
@@ -120,7 +120,8 @@ public void testNonLowercaseNamespace()
                     false,
-                    _ -> false);
+                    _ -> false,
+                    8);
```
Use `new IcebergConfig().getMetadataParallelism()`. Same for others.

ebyhr commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -430,6 +439,16 @@ public IcebergMetadata(
         this.addFilesProcedureEnabled = addFilesProcedureEnabled;
         this.allowedExtraProperties = requireNonNull(allowedExtraProperties, "allowedExtraProperties is null");
+
+        if (metadataParallelism == 1) {
+            this.executorService = directExecutor();
+        }
+        else {
+            this.executorService = Executors.newFixedThreadPool(metadataParallelism, new ThreadFactoryBuilder()
+                    .setNameFormat("iceberg-metadata-pool-%d")
+                    .setDaemon(true) // Optional: Makes threads daemon threads.
+                    .build());
```
Why is the logic different from another metadata executor for listing tables?

shohamyamin replied:
I was follow this where in recent or they changed it to a similar thing https://github.com/trinodb/trino/blob/4718011a39f700ab3c8dbe66747c37a9398922f8/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadataFactory.java#L111

Can you give me an example for another metadata executer so I will follow the same logic

And the use of ThreadFactoryBuilder is required because of the maven checks tests that require that each thread will have a name format

shohamyamin commented:
@ebyhr I have change the code to follow the way that this kind of things are done I was taking inspiration from that  [PR](https://github.com/trinodb/trino/pull/24110) that is very similar and I use the already existing `createIcebergMetadataExecutor` that is located in `IcebergModule`

ebyhr commented:
Could you rebase on master to resolve conflicts?

shohamyamin commented:
> Could you rebase on master to resolve conflicts?

@ebyhr I’ve rebased as requested, and the PR is now updated and ready for review again. Let me know if there’s anything else I should address.

ebyhr commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -423,7 +428,8 @@ public IcebergMetadata(
             boolean addFilesProcedureEnabled,
             Predicate<String> allowedExtraProperties,
-            ExecutorService executor)
+            ExecutorService executor,
+            Executor metadataFetchingExecutor)
```
Now `executor` is too general. Please rename to `systemTableExecutor` or something.

ebyhr commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataFactory.java:
```diff
@@ -53,7 +57,8 @@ public IcebergMetadataFactory(
             @ForIcebergScanPlanning ExecutorService executor,
-            IcebergConfig config)
+            IcebergConfig config,
+            @ForIcebergMetadata ExecutorService metadataExecutorService)
```
Replace the order between `config` and `metadataExecutorService`.

ebyhr commented:
> Improve Performance for Retrieving Tables Metadata from Iceberg Catalog

You can change the commit title to "Improve performance when listing columns in Iceberg".

ebyhr commented:
/test-with-secrets sha=971fa03cb8d4113057825347477eee3cabe06586

piotrrzysko commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -981,23 +987,30 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
+                                try {
+                                    Table icebergTable = catalog.loadTable(session, tableName);
+                                    List<ColumnMetadata> columns = getColumnMetadatas(icebergTable.schema(), typeManager);
+                                    synchronized (tableMetadatas) {
+                                        tableMetadatas.add(TableColumnsMetadata.forTable(tableName, columns));
+                                    }
```
I think this synchronized block can be avoided by returning a result (`TableColumnsMetadata`) from each task.

piotrrzysko commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -981,23 +986,30 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
+                            }, metadataFetchingExecutor))
+                            .toList();
+
+                    CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
```
How will this behave if the query is cancelled? Will pending tasks also be cancelled? I’d suggest using `ExecutorUtil#processWithAdditionalThreads` instead, as it handles such scenarios.

shohamyamin replied:
I have updated the code to use`processWithAdditionalThreads `

raunaqmorarka commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataFactory.java:
```diff
@@ -52,7 +52,7 @@ public IcebergMetadataFactory(
             TableStatisticsWriter tableStatisticsWriter,
             @RawHiveMetastoreFactory Optional<HiveMetastoreFactory> metastoreFactory,
-            @ForIcebergScanPlanning ExecutorService executor,
+            @ForIcebergScanPlanning ExecutorService systemTableExecutor,
```
This executor is not specific to system tables, it is used for planning any iceberg scan which could be system table or not.
Either we shouldn't rename it or rename to icebergScanExecutor

shohamyamin replied:
I've renamed it to `icebergScanExecutor` as you suggested. Keeping it as just `executor` could cause confusion, especially since I'm adding a new executor named `metadataFetchingExecutor`.

raunaqmorarka commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -981,23 +989,40 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
+
+                    try {
+                        List<TableColumnsMetadata> taskResults = processWithAdditionalThreads(tasks, metadataFetchingExecutor).stream()
+                                .filter(Objects::nonNull) // Filter out null results
```
Use `Optional#empty` instead of `null`

shohamyamin commented:
@raunaqmorarka I update the code as requested
